### ISSUE TITLE
MSRV-dependent dependency version resolution

### ISSUE THREAD

**Issue by newpavlov** [rust-lang/rust#89136]
Based on [RFC 2495](https://github.com/rust-lang/rfcs/pull/2495) Rust now supports the [`rust-version`](https://doc.rust-lang.org/nightly/cargo/reference/manifest.html#the-rust-version-field) field, which currently works as:

>If the currently selected version of the Rust compiler is older than the stated version, cargo will exit with an error, telling the user what version is require

This behavior is not ideal. If a project depends on crate `foo v0.1.0` with MSRV 1.56, then releasing crate `foo v0.1.1` with MSRV 1.60 will break build of the project on older toolchains after simple `cargo update`. Ideally Cargo would select `foo v0.1.0` on older tolchains, thus preventing the breakage.

RFC 2495 has [described](https://github.com/rust-lang/rfcs/blob/master/text/2495-min-rust-version.md#influencing-version-resolution) MSRV-dependent version resolution as a potential future extension. This issue is intended for design discussions and tracking implementation progress of this feature.

**Comment by djc** [comment]
I think this probably makes more sense in the Cargo repository, since version resolution happens in Cargo -- might even make sense to do an RFC for it. @ehuss @Eh2406 what do you think?

I can see the attraction of this feature, but I'm also worried about it -- if this behavior doesn't come with a warning, it's trivially possible that downstream crates no longer get updated to upstream dependencies that fixed security vulnerabilities, for example. While I usually deploy cargo-deny in CI to make sure I'm aware of this, as long as something similar is not integrated in Cargo it would be a little scary that semver-compatible updates no longer flow as freely as they used to.

**Comment by newpavlov** [comment]
Yes, having a Cargo warning for cases when there are dependency updates with incompatible MSRV would be a good feature.

As for security vulnerabilities, I think it should be solved by advisories like [RustSec](https://rustsec.org/) and by yanking versions with security vulnerabilities. Also note that today crates can have minor (for pre-1.0, major otherwise) version updates with security fixes which do not get ported to older releases. So i don't think that the proposed MSRV-dependent version resolution is fundamentally different from the existing status quo.
